connection running loop panic when drop column

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
drop column repeatly

<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)
no connection running loop panic

### 3. What did you see instead (Required)
connection running loop panic when drop column
[2022/09/17 21:59:50.346 +00:00] [ERROR] [conn.go:1044] ["connection running loop panic"] [conn=801159147578985855] [lastSQL="DELETE FROM sbtest26 WHERE id=? [arguments: 5007391]"] [err="runtime error: invalid memory address or nil pointer dereference"] [stack="[github.com/pingcap/tidb/server.(*clientConn).Run.func1](http://github.com/pingcap/tidb/server.(*clientConn).Run.func1)\n\t/home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tidb/server/conn.go:1047](http://github.com/pingcap/tidb/server/conn.go:1047)\nruntime.gopanic\n\t/usr/local/go/src/runtime/panic.go:884\nruntime.panicmem\n\t/usr/local/go/src/runtime/panic.go:260\nruntime.sigpanic\n\t/usr/local/go/src/runtime/signal_unix.go:835\ncontainer/list.(*List).Remove\n\t/usr/local/go/src/container/list/list.go:135\[ngithub.com/pingcap/tidb/planner/core.(*LRUPlanCache).removeOldest](http://ngithub.com/pingcap/tidb/planner/core.(*LRUPlanCache).removeOldest)\n\t/home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tidb/planner/core/plan_cache_lru.go:176](http://github.com/pingcap/tidb/planner/core/plan_cache_lru.go:176)\[ngithub.com/pingcap/tidb/planner/core.(*LRUPlanCache).memoryControl](http://ngithub.com/pingcap/tidb/planner/core.(*LRUPlanCache).memoryControl)\n\t/home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tidb/planner/core/plan_cache_lru.go:195](http://github.com/pingcap/tidb/planner/core/plan_cache_lru.go:195)\[ngithub.com/pingcap/tidb/planner/core.(*LRUPlanCache).Put](http://ngithub.com/pingcap/tidb/planner/core.(*LRUPlanCache).Put)\n\t/home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tidb/planner/core/plan_cache_lru.go:115](http://github.com/pingcap/tidb/planner/core/plan_cache_lru.go:115)\[ngithub.com/pingcap/tidb/planner/core.generateNewPlan](http://ngithub.com/pingcap/tidb/planner/core.generateNewPlan)\n\t/home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tidb/planner/core/plan_cache.go:294](http://github.com/pingcap/tidb/planner/core/plan_cache.go:294)\[ngithub.com/pingcap/tidb/planner/core.GetPlanFromSessionPlanCache](http://ngithub.com/pingcap/tidb/planner/core.GetPlanFromSessionPlanCache)\n\t/home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tidb/planner/core/plan_cache.go:165](http://github.com/pingcap/tidb/planner/core/plan_cache.go:165)\[ngithub.com/pingcap/tidb/planner.OptimizeExecStmt](http://ngithub.com/pingcap/tidb/planner.OptimizeExecStmt)\n\t/home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tidb/planner/optimize.go:417](http://github.com/pingcap/tidb/planner/optimize.go:417)\[ngithub.com/pingcap/tidb/planner.Optimize](http://ngithub.com/pingcap/tidb/planner.Optimize)\n\t/home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tidb/planner/optimize.go:129](http://github.com/pingcap/tidb/planner/optimize.go:129)\[ngithub.com/pingcap/tidb/executor.(*Compiler).Compile](http://ngithub.com/pingcap/tidb/executor.(*Compiler).Compile)\n\t/home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tidb/executor/compiler.go:96](http://github.com/pingcap/tidb/executor/compiler.go:96)\[ngithub.com/pingcap/tidb/session.(*session).ExecuteStmt](http://ngithub.com/pingcap/tidb/session.(*session).ExecuteStmt)\n\t/home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tidb/session/session.go:2017](http://github.com/pingcap/tidb/session/session.go:2017)\[ngithub.com/pingcap/tidb/server.(*TiDBContext).ExecuteStmt](http://ngithub.com/pingcap/tidb/server.(*TiDBContext).ExecuteStmt)\n\t/home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tidb/server/driver_tidb.go:231](http://github.com/pingcap/tidb/server/driver_tidb.go:231)\[ngithub.com/pingcap/tidb/server.(*clientConn).executePreparedStmtAndWriteResult](http://ngithub.com/pingcap/tidb/server.(*clientConn).executePreparedStmtAndWriteResult)\n\t/home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tidb/server/conn_stmt.go:250](http://github.com/pingcap/tidb/server/conn_stmt.go:250)\[ngithub.com/pingcap/tidb/server.(*clientConn).executePlanCacheStmt](http://ngithub.com/pingcap/tidb/server.(*clientConn).executePlanCacheStmt)\n\t/home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tidb/server/conn_stmt.go:209](http://github.com/pingcap/tidb/server/conn_stmt.go:209)\[ngithub.com/pingcap/tidb/server.(*clientConn).handleStmtExecute](http://ngithub.com/pingcap/tidb/server.(*clientConn).handleStmtExecute)\n\t/home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tidb/server/conn_stmt.go:203](http://github.com/pingcap/tidb/server/conn_stmt.go:203)\[ngithub.com/pingcap/tidb/server.(*clientConn).dispatch](http://ngithub.com/pingcap/tidb/server.(*clientConn).dispatch)\n\t/home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tidb/server/conn.go:1386](http://github.com/pingcap/tidb/server/conn.go:1386)\[ngithub.com/pingcap/tidb/server.(*clientConn).Run](http://ngithub.com/pingcap/tidb/server.(*clientConn).Run)\n\t/home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tidb/server/conn.go:1109](http://github.com/pingcap/tidb/server/conn.go:1109)\[ngithub.com/pingcap/tidb/server.(*Server).onConn](http://ngithub.com/pingcap/tidb/server.(*Server).onConn)\n\t/home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tidb/server/server.go:563](http://github.com/pingcap/tidb/server/server.go:563)"]

### 4. What is your TiDB version? (Required)
./tidb-server -V
 Release Version: v6.3.0-alpha
Edition: Community
Git Commit Hash: efc07201e3441ed8dfed08e10ae19fa407dda596
Git Branch: heads/refs/tags/v6.3.0-alpha
UTC Build Time: 2022-09-17 11:08:40
GoVersion: go1.19
Race Enabled: false
TiKV Min Version: 6.2.0-alpha
Check Table Before Drop: false

<!-- Paste the output of SELECT tidb_version() -->

--- a/planner/core/plan_cache_lru.go
+++ b/planner/core/plan_cache_lru.go
@@ -169,6 +169,9 @@ func (l *LRUPlanCache) SetCapacity(capacity uint) error {
 // removeOldest removes the oldest element from the cache.
 func (l *LRUPlanCache) removeOldest() {
 	lru := l.lruList.Back()
+	if lru == nil {
+		return
+	}
 	if l.onEvict != nil {
 		l.onEvict(lru.Value.(*planCacheEntry).PlanKey, lru.Value.(*planCacheEntry).PlanValue)
 	}

--- a/planner/core/plan_cache_lru.go
+++ b/planner/core/plan_cache_lru.go
@@ -191,7 +191,7 @@ func (l *LRUPlanCache) memoryControl() {
 	}
 
 	memUsed, _ := memory.InstanceMemUsed()
-	for memUsed > uint64(float64(l.quota)*(1.0-l.guard)) {
+	for memUsed > uint64(float64(l.quota)*(1.0-l.guard)) && l.size > 0 {
 		l.removeOldest()
 		memUsed, _ = memory.InstanceMemUsed()
 	}

--- a/planner/core/plan_cache_lru_test.go
+++ b/planner/core/plan_cache_lru_test.go
@@ -330,5 +330,7 @@ func TestIssue37914(t *testing.T) {
 		tps:  pTypes,
 	}
 
-	lru.Put(key, val, pTypes)
+	require.NotPanics(t, func() {
+		lru.Put(key, val, pTypes)
+	})
 }

--- a/planner/core/plan_cache_lru_test.go
+++ b/planner/core/plan_cache_lru_test.go
@@ -319,3 +319,16 @@ func TestLRUPCSetCapacity(t *testing.T) {
 	err = lru.SetCapacity(0)
 	require.Error(t, err, "capacity of LRU cache should be at least 1")
 }
+
+func TestIssue37914(t *testing.T) {
+	lru := NewLRUPlanCache(3, 0.1, 1, pickFromBucket)
+
+	pTypes := []*types.FieldType{types.NewFieldType(mysql.TypeFloat), types.NewFieldType(mysql.TypeDouble)}
+	key := newMockHashKey(int64(1))
+	val := &fakePlan{
+		plan: int64(1),
+		tps:  pTypes,
+	}
+
+	lru.Put(key, val, pTypes)
+}

